src/librustc_parse/parser/item.rs

```diff
@@ -212,9 +216,13 @@ impl<'a> Parser<'a> {
         } else if vis.node.is_pub() && self.isnt_macro_invocation() {
             self.recover_missing_kw_before_item()?;
             return Ok(None);
-        } else if macros_allowed && self.token.is_path_start() {
+        } else if let Some(kind) = if macros_allowed && self.token.is_path_start() {
+            self.parse_item_macro(vis, mod_stmt)?
+        } else {
+            None
+        } {
             // MACRO INVOCATION ITEM
-            (Ident::invalid(), ItemKind::Mac(self.parse_item_macro(vis)?))
+            (Ident::invalid(), ItemKind::Mac(kind))
         } else {
```

Comment:
@estebank we could do something like:

```suggestion
        } else if self.maybe_consume_incorrect_semicolon(&[]) {
            let lo = self.token.span;
            self.parse_item_kind(attrs, macros_allowed, lo, vis, def, req_name, mod_stmt)?
        } else {
```